lib/luminork-server/src/search/parser.rs

```diff
@@ -0,0 +1,555 @@
+            // unclosed quotes should still show the partial match!
+            SearchTerm::StartsWith(value)
+        }
+    }
+
+    /// Parse the attribute values with | or , immediately between them, as part of
+    /// attr:value1|value2|value3|...
+    ///
+    /// Consumes until &/space, (, ), !, or end of string
+    fn parse_attr_terms(&mut self) -> Vec<SearchTerm> {
+        let mut values = vec![];
+        loop {
+            if let Some(value) = self.parse_attr_value() {
+                values.push(value);
+            }
+
+            // If there's a | or ,, continue parsing more values
+            if !(self.consume('|') || self.consume(',')) {
+                break;
+            }
+        }
+        values
+    }
+
+    /// Parse a single attribute value alternative, like value1 in attr:value1|value2|value3|...
+    ///
+    /// Consumes until &/space, |/,, (, ), !, or end of string
+    ///
+    /// Returns None if we're at &/space, |/,, ), or end of string
+    fn parse_attr_value(&mut self) -> Option<SearchTerm> {
+        // If the string starts with ", we treat it as an exact match value (unless it's not closed, read below)
+        if self.consume('"') {
+            let value = self.consume_until(['"']).to_string();
+            // Consume the closing quote so we can move on
+            if self.consume('"') {
+                Some(SearchTerm::Exact(value))
+            } else {
+                // If there's no close quote, treat it as a startsWith match to improve UX while typing
+                Some(SearchTerm::StartsWith(value))
+            }
+        } else {
+            // No quotes, it's a normal match for the given value
+            let value = self.consume_until([' ', '(', ')', '&', '!', '"', '|', ',']);
+            if value.is_empty() {
+                None
+            } else {
+                Some(SearchTerm::Match(value.to_string()))
+            }
+        }
+    }
+
+    /// Consume the given char if it's next in the string, returning true if it was.
+    fn consume(&mut self, ch: char) -> bool {
+        match self.remaining_query_string.strip_prefix(ch) {
+            Some(remaining) => {
+                self.remaining_query_string = remaining;
+                true
+            }
+            None => false,
+        }
+    }
+
+    /// Consume all characters until you reach one of the given chars, or end of string.
+    /// Does not consume the found char.
+    fn consume_until<const N: usize>(&mut self, chars: [char; N]) -> &str {
+        let index = self
+            .remaining_query_string
+            .find(chars)
+            .unwrap_or(self.remaining_query_string.len());
+        let (consumed, remaining) = self.remaining_query_string.split_at(index);
+        self.remaining_query_string = remaining;
+        consumed
+    }
+
+    /// Whether we're at the end of the query string.
+    fn finished(&self) -> bool {
+        self.remaining_query_string.is_empty()
+    }
+}
+
+#[cfg(test)]
```

Comment:
Unit tests for anything that can be reasonably tested! In this case, we try to test all the syntax variants we can think of and make sure they produce the right structure. These tests do *not* test whether the parsed query is actually processed correctly to match components, but this handles a lot of complex variants so that the downstream component matching tests can be simpler.